src/dmd/dsymbolsem.d

```diff
@@ -755,8 +755,15 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
         if (dsym.semanticRun >= PASS.semanticdone)
             return;
 
-        if (sc && sc.inunion && sc.inunion.isAnonDeclaration())
-            dsym.overlapped = true;
+        if (sc && sc.inunion)
+        {
+            auto anonDecl = sc.inunion.isAnonDeclaration();
+            if (anonDecl)
+            {
+                dsym.overlapped = true;
+                anonDecl.fields.push(dsym);
+            }
+        }
```

Comment:
Using sc.inunion will fail because it doesn't take into account nesting:
```
struct S {
    union {
      struct {
         int a = 3;
         int b = 4;  <= this should work
     }
   }
}
```
That's why the rest of the code checks for "overlapping".

A correct algorithm would check for overlapping initializations, which is the real problem.

Reply:
With my patch your example does work. `sc.isunion` has been modified; it is not a simple bool anymore, it is a dsymbol which can be either a union declaration or an anon declaration; in this situation, when b is analyzed, the code that I've written sees that it is actually inside a struct scope.

This is the right algorithm. We want to issue an error for initializations of other members than the first. Please reconsider.

Reply:
How about this:
```
struct S {
    union {
      struct {
         int a = 3;
         int b = 4;  <= this should work
     }
      struct {
         int c = 3;
         int d = 4;  <= this should fail
     }
   }
}
```
and this:
```
struct S {
    union {
      struct {
         union { int a = 3; }
        }
       int b = 4;  <= this should fail
     }
   }
}
```
These should go in the test cases.

Reply:
All this stuff is already taken care of in the code that generates the initializer for the struct. It just has to be improved to give an error on an overlapping explicit initializer.

Reply:
(current behavior is to just skip it)

Reply:
This patch does not affect the above code snippets. This patch is about initializing a field that is different than the first field. Up until now, dmd silently ignored the initializer in such situations, while now it will issue a deprecation highlighting the fact that the provided initializer will not be used:

```d
union U
{
    int n;
    float f = 3.14f;
}
U u;
assert(u.f == 3.14f);
```

Without this patch, this code compiles and the assertion fails (int.init is used to initialize the union), which is surprising. With this patch, the code will issue a deprecation that attentions the user that the initializer he provided will be ignored. This patch is not about overlapping initialization, it does not modify that; this patch takes care of initializations of fields that are not the first field of a union; all else is exactly the same as it used to be.

Reply:
I think this patch doesn't account for nesting of anonymous structs and unions. There are no tests added for those cases.

Reply:
It seems to me the problem is in aggregate.d in the functions fill() and checkOverlappedFields(), and that's where the solution should be.

Reply:
I added the snippets you provided as tests.